crates/environ/src/fact/trampoline.rs

```diff
@@ -80,50 +85,168 @@ struct Compiler<'a, 'b> {
+        let result = module
+            .funcs
+            .push(Function::new(Some(adapter.name.clone()), ty));
+
+        // If this type signature contains any borrowed resources then invocations
+        // of enter/exit call for resource-related metadata tracking must be used.
+        // It shouldn't matter whether the lower/lift signature is used here as both
+        // should return the same answer.
+        let emit_resource_call = module.types.contains_borrow_resource(&adapter.lower);
+        assert_eq!(
+            emit_resource_call,
+            module.types.contains_borrow_resource(&adapter.lift)
+        );
+
+        (
+            Compiler {
+                types: module.types,
+                module,
+                code: Vec::new(),
+                nlocals: lower_sig.params.len() as u32,
+                free_locals: HashMap::new(),
+                traps: Vec::new(),
+                result,
+                fuel: INITIAL_FUEL,
+                emit_resource_call,
+            },
+            lower_sig,
+            lift_sig,
+        )
+    }
+
+    let start_adapter = |module: &mut Module, param_globals| {
+        let sig = module.types.async_start_signature(&adapter.lift);
+        let ty = module.core_types.function(&sig.params, &sig.results);
+        let result = module.funcs.push(Function::new(
+            Some(format!("[async-start]{}", adapter.name)),
+            ty,
+        ));
+
+        Compiler {
+            types: module.types,
+            module,
+            code: Vec::new(),
+            nlocals: sig.params.len() as u32,
+            free_locals: HashMap::new(),
+            traps: Vec::new(),
+            result,
+            fuel: INITIAL_FUEL,
+            emit_resource_call: false,
+        }
+        .compile_async_start_adapter(adapter, &sig, param_globals);
+
+        result
+    };
+
+    let return_adapter = |module: &mut Module, result_globals| {
+        let sig = module.types.async_return_signature(&adapter.lift);
+        let ty = module.core_types.function(&sig.params, &sig.results);
+        let result = module.funcs.push(Function::new(
+            Some(format!("[async-return]{}", adapter.name)),
+            ty,
+        ));
+
+        Compiler {
+            types: module.types,
+            module,
+            code: Vec::new(),
+            nlocals: sig.params.len() as u32,
+            free_locals: HashMap::new(),
+            traps: Vec::new(),
+            result,
+            fuel: INITIAL_FUEL,
+            emit_resource_call: false,
+        }
+        .compile_async_return_adapter(adapter, &sig, result_globals);
+
+        result
+    };
+
     match (adapter.lower.options.async_, adapter.lift.options.async_) {
-        (false, false) => {}
+        (false, false) => {
+            let (compiler, lower_sig, lift_sig) = compiler(module, adapter);
+            compiler.compile_sync_to_sync_adapter(adapter, &lower_sig, &lift_sig)
+        }
         (true, true) => {
```

Comment:
Mind leaving some high-level comments on each of these cases about what sort of adapters are being created? Mostly to help guide what adapters are calling what and what the purpose of the various adapter functions are.

Reply:
(I'm also not sure at-a-glance, for example, what the `param_globals`/`result_globals` variables are, so either docs here or on the methods below I think would be useful)

Reply:
I just pushed an update that adds some comments.  Let me know if anything needs to be clarified or added.